org.openscenegraph.android/jni/JNIosg.cpp

```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+	osg::Vec3Array* a = reinterpret_cast<osg::Vec3Array*>(array_cptr);
+
+	if( (g!=NULL) && (a!=NULL) )
+	{
+		g->setVertexArray(a);
+	}
+}
+
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeGetVertexArray(JNIEnv* env, jclass, jlong cptr)
+{
+	jlong result = 0;
+	osg::Geometry* g = reinterpret_cast<osg::Geometry*>(cptr);
+	if(g!=NULL)
+	{
+		osg::Vec3Array* a = reinterpret_cast<osg::Vec3Array*>(g->getVertexArray());
+		result = reinterpret_cast<jlong>(a);
+	}
+	return result;
+}
+
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeSetNormalArray(JNIEnv* env, jclass, jlong cptr, jobjectArray inputBuffer)
+{
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+
+	int array_n = env->GetArrayLength(inputBuffer);
+	osg::Vec3Array* _array = new osg::Vec3Array();
+
+	jobject objArray;
+	jsize subarray_n;
+	float* data;
+	for(int i = 0; i < array_n; i++)
+	{
+		objArray = env->GetObjectArrayElement(inputBuffer,i);
+		jfloatArray* cdata = reinterpret_cast<jfloatArray*>(&objArray);
+		subarray_n = env->GetArrayLength(*cdata);
+		if( subarray_n>2 )
+		{
+			data = env->GetFloatArrayElements(*cdata, 0);
+			_array->push_back(osg::Vec3(data[0] , data[1], data[2]));
+			env->ReleaseFloatArrayElements(*cdata, data, 0);
+		}
+	}
+
+	if(g!=NULL)
+	{
+		g->setNormalArray(_array);
+	}
+}
+
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeSetNormalArrayNative(JNIEnv* env, jclass, jlong cptr, jlong array_cptr)
+{
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+	osg::Vec3Array* a = reinterpret_cast<osg::Vec3Array*>(array_cptr);
+
+	if( (g!=NULL) && (a!=NULL) )
+	{
+		g->setNormalArray(a);
+	}
+}
+
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeGetNormalArray(JNIEnv* env, jclass, jlong cptr)
+{
+	jlong result = 0;
+	osg::Geometry* g = reinterpret_cast<osg::Geometry*>(cptr);
+	if(g!=NULL)
+	{
+		osg::Vec3Array* a = reinterpret_cast<osg::Vec3Array*>(g->getNormalArray());
+		result = reinterpret_cast<jlong>(a);
+	}
+	return result;
+}
+
+JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Geometry_nativeSetColorArray(JNIEnv* env, jclass, jlong cptr, jobjectArray inputBuffer)
+{
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+
+	int array_n = env->GetArrayLength(inputBuffer);
+
+	//LOGI("Colour Array size: %i\n", array_n);
+	osg::Vec4Array* _array = new osg::Vec4Array();
+
+	jobject objArray;
+	jsize subarray_n;
+	float* data;
+	for(int i = 0; i < array_n; i++)
+	{
+		objArray = env->GetObjectArrayElement(inputBuffer,i);
+		jfloatArray* cdata = reinterpret_cast<jfloatArray*>(&objArray);
+		subarray_n = env->GetArrayLength(*cdata);
+		//LOGI("subarray size at %i: %i\n", i, subarray_n);
+		if( subarray_n>3 )
+		{
+			data = env->GetFloatArrayElements(*cdata, 0);
+			_array->push_back(osg::Vec4(data[0], data[1], data[2], data[3]));
+			env->ReleaseFloatArrayElements(*cdata, data, 0);
+		}
+	}
+
+	if(g!=NULL)
```

Comment:
agreed. I will go over it again and adapt late terminations towards early terminations.